### ISSUE TITLE
XML - Remove empty tags

### ISSUE THREAD

**Issue by yeikel** [openrewrite/rewrite#2737]
Provide a generic recipe to remove empty tags from a given XML file. 

This could be useful for changes like #2736 where the recipe leaves the pom file with an empty block that is no longer needed

Example : 

```xml
<project>
<groupId>com.mycompany.app</groupId>
<artifactId>my-app</artifactId>
<version>1</version>
<pluginRepositories>
</pluginRepositories>
</project>
```

```xml
<project>
<groupId>com.mycompany.app</groupId>
<artifactId>my-app</artifactId>
<version>1</version>
</project>
```

**Comment by timtebeek** [comment]
Agree that this would be a nice addition as a tag-on recipe after dropping other xml elements. For anyone looking to implement this: be sure to also include a test case where a tag is otherwise empty, but does have attributes that should prevent it from being removed.

**Comment by knutwannheden** [comment]
I wonder if this is really a good idea. In many XML files empty elements also have semantics. E.g. XHTML's `<br/>` or Maven's `<relativePath>`. In both these examples there are of course more cases when an empty element can be removed.

I think that if it isn't already trivial to remove an empty element explicitly using a programmatic recipe, we could add some helper method. But I am unsure about a general purpose XML recipe to remove all empty tags.

**Comment by knutwannheden** [comment]
Yes, I suppose some kind of `RemoveTag` visitor which has the option to also remove any resulting empty ancestor elements would make sense.

**Comment by vishalsingh2972** [comment]
@yeikel could you assign this to me if it's still open?

**Comment by yeikel** [comment]
I don't have permission to assign, but @timtebeek can

**Comment by timtebeek** [comment]
Much appreciated @vishalsingh2972 ; Let us know how you do, and feel free to swing by in Slack for any help as well.

**Comment by timtebeek** [comment]
Once implemented this should also be added after `RemoveProperty` as per https://github.com/openrewrite/rewrite/issues/3231.

**Comment by timtebeek** [comment]
Hope you don't mind I unassigned you @vishalsingh2972 
I want to open this issue back up again to anyone to pick up.
Feel free to open up a draft pull request if you'd like to contribute!

**Comment by TheMarvelFan** [comment]
Hi is anyone working on this issue?

**Comment by knutwannheden** [comment]
@TheMarvelFan At the moment nobody is working on this.

**Comment by TheMarvelFan** [comment]
I would like to work on it, but I will need some guidance.

**Comment by timtebeek** [comment]
Hi @TheMarvelFan ! Thanks for the offer to help; I think it might be best to start with a draft pull request that outlines the cases that we want to cover as individual unit tests. I'm thinking of cases such as the following.

Do removed when there are no child elements, regardless of style.
```xml
<pluginRepositories />
```
```xml
<pluginRepositories>
</pluginRepositories>
```
But not remove when there are attributes, or child elements.
```xml
<pluginRepositories some="attribute">
</pluginRepositories>
```

You can look towards the existing xml unit tests we already have, and refine once you have a start to work from:
https://github.com/openrewrite/rewrite/tree/main/rewrite-xml/src/test/java/org/openrewrite/xml

**Comment by TheMarvelFan** [comment]
@timtebeek Hi I have added a draft pull request (#3613), and I will keep adding commits to it till it is ready to merge. Please feel free to review and request for changes.

**Comment by timtebeek** [comment]
Awesome! Great to see you get started! I'll keep an eye on your pull request. Let us know if you need any help!
